File: app/src/main/java/jp/hazuki/yuzubrowser/utils/view/recycler/ArrayRecyclerAdapter.java
/*
 * Copyright (C) 2017 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.utils.view.recycler;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.util.SparseBooleanArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;


public abstract class ArrayRecyclerAdapter<T, VH extends ArrayRecyclerAdapter.ArrayViewHolder<T>> extends RecyclerView.Adapter<VH> {

    private List<T> items;
    private OnRecyclerListener recyclerListener;
    private LayoutInflater inflater;
    private boolean sortMode;

    private boolean multiSelectMode;
    private SparseBooleanArray itemSelected;

    public ArrayRecyclerAdapter(Context context, List<T> list, OnRecyclerListener listener) {
        itemSelected = new SparseBooleanArray();
        items = list;
        recyclerListener = listener;
        inflater = LayoutInflater.from(context);
        sortMode = false;
    }

    public void onBindViewHolder(VH holder, T item, int position) {
    }

    protected abstract VH onCreateViewHolder(LayoutInflater inflater, ViewGroup parent, int viewType);

    public void move(int fromPosition, int toPosition) {
        if (fromPosition < toPosition) {
            for (int i = fromPosition; i < toPosition; i++) {
                Collections.swap(items, i, i + 1);
            }
        } else {
            for (int i = fromPosition; i > toPosition; i--) {
                Collections.swap(items, i, i - 1);
            }
        }
        notifyItemMoved(fromPosition, toPosition);
    }

    @Override
    public int getItemCount() {
        if (items != null)
            return items.size();
        else
            return 0;
    }

    public List<T> getItems() {
        return items;
    }

    public int size() {
        return items.size();
    }

    public T get(int index) {
        return items.get(index);
    }

    public void add(T item) {
        items.add(item);
    }

    public void add(int index, T item) {
        items.add(index, item);
    }

    public void addAll(Collection<T> collections) {
        items.addAll(collections);
    }

    public T set(int index, T item) {
        return items.set(index, item);
    }

    public T remove(int index) {
        T item = items.remove(index);
        notifyItemRemoved(index);
        return item;
    }

    public boolean remove(T item) {
        int index = items.indexOf(item);
        if (index >= 0) {
            remove(index);
            return true;
        } else {
            return false;
        }
    }

    public void clear() {
        items.clear();
    }

    public OnRecyclerListener getListener() {
        return recyclerListener;
    }

    public boolean isSortMode() {
        return sortMode;
    }

    public void setSortMode(boolean sort) {
        if (sort != sortMode) {
            sortMode = sort;
            notifyDataSetChanged();
        }
    }

    protected void setRecyclerListener(OnRecyclerListener listener) {
        recyclerListener = listener;
    }

    @Override
    public VH onCreateViewHolder(ViewGroup parent, int viewType) {
        return onCreateViewHolder(inflater, parent, viewType);
    }

    @Override
    public void onBindViewHolder(final VH holder, int position) {
        int pos = holder.getAdapterPosition();
        if (items != null && items.size() > pos && items.get(pos) != null) {
            holder.setUp(items.get(pos));
            onBindViewHolder(holder, items.get(pos), pos);
        }
    }

    private void onItemClick(View v, int position, T item) {
        position = searchPosition(position, item);
        if (position < 0) return;
        if (multiSelectMode) {
            toggle(position);
        } else if (recyclerListener != null)
            recyclerListener.onRecyclerItemClicked(v, position);
    }

    private boolean onItemLongClick(View v, int position, T item) {
        position = searchPosition(position, item);
        return position >= 0 && !sortMode && recyclerListener != null && recyclerListener.onRecyclerItemLongClicked(v, position);
    }

    protected int searchPosition(int position, T item) {
        if (position < 0 || position >= getItemCount() || !get(position).equals(item)) {
            if (position > 0 && get(position - 1).equals(item))
                return position - 1;

            position = items.indexOf(item);
            if (position < 0) notifyDataSetChanged();
            return position;
        }
        return position;
    }

    public boolean isMultiSelectMode() {
        return multiSelectMode;
    }

    public void setMultiSelectMode(boolean multiSelect) {
        if (multiSelect != multiSelectMode) {
            multiSelectMode = multiSelect;

            if (!multiSelect) {
                itemSelected.clear();
            }

            notifyDataSetChanged();
        }
    }

    public void toggle(int position) {
        setSelect(position, !itemSelected.get(position, false));
    }

    public void setSelect(int position, boolean isSelect) {
        boolean old = itemSelected.get(position, false);
        itemSelected.put(position, isSelect);

        if (old != isSelect) {
            notifyItemChanged(position);
        }
    }

    public boolean isSelected(int position) {
        return itemSelected.get(position, false);
    }

    public List<Integer> getSelectedItems() {
        List<Integer> items = new ArrayList<>();
        for (int i = 0; itemSelected.size() > i; i++) {
            if (itemSelected.valueAt(i)) {
                items.add(itemSelected.keyAt(i));
            }
        }
        return items;
    }

    public static class ArrayViewHolder<I> extends RecyclerView.ViewHolder {

        private I item;

        public ArrayViewHolder(View itemView, final ArrayRecyclerAdapter<I, ?> adapter) {
            super(itemView);
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    adapter.onItemClick(v, getAdapterPosition(), item);
                }
            });

            itemView.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View v) {
                    return adapter.onItemLongClick(v, getAdapterPosition(), item);
                }
            });
        }

        public void setUp(I item) {
            this.item = item;
        }

        protected I getItem() {
            return item;
        }
    }
}

File: module/ui/src/main/java/jp/hazuki/yuzubrowser/ui/widget/recycler/ArrayRecyclerAdapter.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.ui.widget.recycler

import android.content.Context
import android.util.SparseBooleanArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import java.util.*


abstract class ArrayRecyclerAdapter<T, VH : ArrayRecyclerAdapter.ArrayViewHolder<T>>(
        context: Context,
        val items: MutableList<T>,
        private var listener: OnRecyclerListener?
) : RecyclerView.Adapter<VH>() {
    private val inflater: LayoutInflater = LayoutInflater.from(context)
    private var sortMode: Boolean = false

    var isMultiSelectMode: Boolean = false
        set(multiSelect) {
            if (multiSelect != isMultiSelectMode) {
                field = multiSelect

                if (!multiSelect) {
                    itemSelected.clear()
                    selectedItemCount = 0
                }

                notifyDataSetChanged()
            }
        }
    private val itemSelected: SparseBooleanArray = SparseBooleanArray()

    var selectedItemCount: Int = 0
        private set

    var isSortMode: Boolean
        get() = sortMode
        set(sort) {
            if (sort != sortMode) {
                sortMode = sort
                notifyDataSetChanged()
            }
        }

    val selectedItems: List<Int>
        get() {
            return (0 until itemSelected.size())
                    .asSequence()
                    .filter { itemSelected.valueAt(it) }
                    .map { itemSelected.keyAt(it) }
                    .toList()
        }

    open fun onBindViewHolder(holder: VH, item: T, position: Int) {}

    protected abstract fun onCreateViewHolder(inflater: LayoutInflater, parent: ViewGroup?, viewType: Int): VH

    fun move(fromPosition: Int, toPosition: Int) {
        if (fromPosition < toPosition) {
            for (i in fromPosition until toPosition) {
                Collections.swap(items, i, i + 1)
            }
        } else {
            for (i in fromPosition downTo toPosition + 1) {
                Collections.swap(items, i, i - 1)
            }
        }
        notifyItemMoved(fromPosition, toPosition)
    }

    override fun getItemCount(): Int {
        return items.size
    }

    fun size(): Int {
        return items.size
    }

    open operator fun get(index: Int): T {
        return items[index]
    }

    fun add(item: T) {
        items.add(item)
    }

    fun add(index: Int, item: T) {
        items.add(index, item)
    }

    fun addAll(collections: Collection<T>) {
        items.addAll(collections)
    }

    operator fun set(index: Int, item: T): T {
        return items.set(index, item)
    }

    fun remove(index: Int): T {
        val item = items.removeAt(index)
        notifyItemRemoved(index)
        return item
    }

    fun remove(item: T): Boolean {
        val index = items.indexOf(item)
        return if (index >= 0) {
            remove(index)
            true
        } else {
            false
        }
    }

    fun clear() {
        items.clear()
    }

    protected fun setRecyclerListener(listener: OnRecyclerListener) {
        this.listener = listener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
        return onCreateViewHolder(inflater, parent, viewType)
    }

    override fun onBindViewHolder(holder: VH, position: Int) {
        val pos = holder.adapterPosition
        if (items.size > pos && items[pos] != null) {
            holder.setUp(items[pos])
            onBindViewHolder(holder, items[pos], pos)
        }
    }

    protected open fun onItemClick(v: View, position: Int, item: T?) {
        var calPosition = position
        calPosition = searchPosition(calPosition, item)
        if (calPosition < 0) return
        if (isMultiSelectMode) {
            toggle(calPosition)
        } else {
            listener?.onRecyclerItemClicked(v, calPosition)
        }
    }

    private fun onItemLongClick(v: View, position: Int, item: T?): Boolean {
        var calPosition = position
        calPosition = searchPosition(calPosition, item)
        return calPosition >= 0 && !sortMode && listener?.onRecyclerItemLongClicked(v, calPosition) ?: false
    }

    protected fun searchPosition(position: Int, item: T?): Int {
        var calPosition = position
        if (calPosition < 0 || calPosition >= itemCount || get(calPosition) != item) {
            if (calPosition > 0 && get(calPosition - 1) == item)
                return calPosition - 1

            calPosition = items.indexOf(item)
            if (calPosition < 0) notifyDataSetChanged()
            return calPosition
        }
        return calPosition
    }

    open fun toggle(position: Int) {
        setSelect(position, !itemSelected.get(position, false))
    }

    open fun setSelect(position: Int, isSelect: Boolean) {
        val old = itemSelected.get(position, false)
        itemSelected.put(position, isSelect)

        if (old != isSelect) {
            notifyItemChanged(position)
            if (isSelect) selectedItemCount++ else selectedItemCount--
        }
    }

    fun isSelected(position: Int): Boolean {
        return itemSelected.get(position, false)
    }

    open class ArrayViewHolder<I>(itemView: View, adapter: ArrayRecyclerAdapter<I, *>) : androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView) {

        private var target: I? = null
        protected val item: I
            get() = target ?: throw NullPointerException()

        init {
            itemView.setOnClickListener { v -> adapter.onItemClick(v, adapterPosition, item) }

            itemView.setOnLongClickListener { v -> adapter.onItemLongClick(v, adapterPosition, item) }
        }

        open fun setUp(item: I) {
            target = item
        }
    }
}
